src/complex/Utilities/DataArrayUtilities.hpp

```diff
@@ -904,4 +906,215 @@ class CopyTupleUsingIndexList
+ *   }
+ *   taskRunner.wait();
+ * @code
+ *
+ * @endcode
+ */
+class CopyFromArray
+{
+public:
+  CopyFromArray()
+  {
+  }
+  ~CopyFromArray() = default;
+
+  CopyFromArray(const CopyFromArray&) = default;
+  CopyFromArray(CopyFromArray&&) noexcept = default;
+  CopyFromArray& operator=(const CopyFromArray&) = delete;
+  CopyFromArray& operator=(CopyFromArray&&) noexcept = delete;
+
+  template <bool UseCombine, bool UseAppend>
+  struct TemplateTypeOptions
+  {
+    static inline constexpr bool UsingCombine = UseCombine;
+    static inline constexpr bool UsingAppend = UseAppend;
+  };
+
+  using Combine = TemplateTypeOptions<true, false>;
+  using Append = TemplateTypeOptions<false, true>;
+
+  template <class TemplateTypeOptions = Combine, class ParallelRunnerT, class... ArgsT>
+  static void RunParallel(IArray::ArrayType arrayType, IArray* destArray, ParallelRunnerT&& runner, ArgsT&&... args)
+  {
+    static_assert(!(TemplateTypeOptions::UsingCombine && TemplateTypeOptions::UsingAppend), "Cannot use both Append and Combine");
+    static_assert(!(!TemplateTypeOptions::UsingCombine && !TemplateTypeOptions::UsingAppend), "Cannot have Append and Combine false");
+
+    DataType dataType = DataType::int32;
+    if(arrayType == IArray::ArrayType::NeighborListArray)
+    {
+      dataType = dynamic_cast<INeighborList*>(destArray)->getDataType();
+    }
+    if(arrayType == IArray::ArrayType::DataArray)
+    {
+      dataType = dynamic_cast<IDataArray*>(destArray)->getDataType();
+      if(dataType == DataType::boolean)
+      {
+        if constexpr(TemplateTypeOptions::UsingCombine)
+        {
+          RunCombineBoolAppend(destArray, std::forward<ArgsT>(args)...);
+        }
+        if constexpr(TemplateTypeOptions::UsingAppend)
+        {
+          RunAppendBoolAppend(destArray, std::forward<ArgsT>(args)...);
+        }
+        return;
+      }
+    }
+
+    if constexpr(TemplateTypeOptions::UsingCombine)
+    {
+      ExecuteParallelFunction<CombineArrays, NoBooleanType>(dataType, std::forward<ParallelRunnerT>(runner), arrayType, destArray, std::forward<ArgsT>(args)...);
+    }
+    if constexpr(TemplateTypeOptions::UsingAppend)
+    {
+      ExecuteParallelFunction<AppendArray, NoBooleanType>(dataType, std::forward<ParallelRunnerT>(runner), arrayType, destArray, std::forward<ArgsT>(args)...);
+    }
+  }
+
+private:
+  template <class K>
+  static void AppendData(K* inputArray, K* destArray, usize offset)
+  {
+    const usize numElements = inputArray->getNumberOfTuples() * inputArray->getNumberOfComponents();
+    for(usize i = 0; i < numElements; ++i)
+    {
+      auto value = (*inputArray)[i]; // make sure we are getting a copy not a ref
+      (*destArray)[offset + i] = value;
+    }
+  }
+
+  template <typename T>
+  class AppendArray
+  {
+  public:
+    AppendArray(IArray::ArrayType arrayType, IArray* destCellArray, IArray* inputCellArray, usize tupleOffset)
```

Comment:
Would it make sense for the array type to be inferred from the first array? And I think that all these functions should take in there arrays by reference. They'll still be stored in a pointer, but we can protect the interface a bit more. Also arrays that are only read from should be const.